deprecate CHANGELOG, improve release action, PR template, Contributing doc

twelch opened the pull request:
This is a draft PR that adds more to the release action, open for discussion.  It:

- Automatically marks a new release as the latest.  I noticed that 6.5 was still marked as the current Github release, so the automated action didn't update that.
- Generates release notes automatically from the commit log and add mention to the contributor. (CHANGELOG is deprecated with a message in favor of using release notes).
- Marks the release as draft, so that a human can review, cleaning up the auto-notes, adding breaking change instructions, and press the button to publish the release record.

This PR also:
- refactors Contributing including new release docs
- Simplifies the pull request template

The release steps need to get tested, and corrected as needed, on the next release.

smallsaucepan commented:
Another thing worth noting is that the URLs of github releases are predictable - the URL for 8.2.3 will be https://github.com/Turfjs/turf/releases/tag/v8.2.3 for example

That means when we're preparing for a release, we could put the following at the top of CHANGELOG.md before we even tag the code:

> ### v8.2.3
> Release notes - https://github.com/Turfjs/turf/releases/tag/v8.2.3

If there are already some changes people have been saving to CHANGELOG.md as we go, just cut them into a draft release and replace with the above.

And that's it. We can augment any manually copied notes with the auto generate feature in the release. Minimises duplication, though still keeps a CHANGELOG.md in the project root for people to use as a jumping off point.

With the changelog side of things removed, automating the release creation doesn't really save us much labour. I'm wary of trying to over-automate things. Publishing to NPM is significant / rare enough an event that it probably benefits from a bit of manual preparation, Slack chatter, etc even if the final publishing process itself is largely automated.

mfedderly reviewed (approved):
I like this approach, I don't think we should be manually managing changelogs.
I assume this means that we need more useful PR titles because that's what gets put into the Github changelogs?

twelch commented on CONTRIBUTING.md:
```diff
@@ -95,11 +95,14 @@ it will create a new folder inside `packages` with a simple boilerplate for your
-- Make a PR that contains the file changes from `pnpm lerna version --no-push 7.0.0` (update the version as necessary)
+- Review code commits and decide what type of release this will be - PATCH, MINOR, MAJOR
+- If necessary, make and merge a PR to contain any last minute housekeeping items (docs, tests, etc.)
+- Make a PR that contains the file changes from `pnpm lerna version --no-push 7.0.0` (update this version number as appropriate)
```
@mfedderly what do you do exactly to "Make a PR" that contains the file changes from the version command?  Do you do this using Github CI? Through the Github web UI?

And how exactly do you "push the release PR" in the follow on step below?

mfedderly replied:
When you run that lerna command it changes the package.json files to update all of the versions everywhere. You just add all the changes and commit it.

twelch replied:
Thanks @mfedderly, the part I didn’t understand is how/where the PR should get created if the next step is to “push the release PR”.  I’m not sure what it means to push a PR. Can you provide more detail on what you type for each step and where you are doing it? (git client, github website).

mfedderly replied:
ah ok so in the console its something like this:

```
git fetch origin
get checkout origin/master -b mf/release-7.0.0
pnpm lerna version --no-push 7.0.0
git add packages/*/package.json (or something like this? check what files got changed)
git commit -m "Release v7.0.0"
git push origin mf/release-7.0.0 --follow-tags
# use the link in the output from ^ to create a PR and get it merged in the normal fashion
```

twelch commented on CONTRIBUTING.md:
```diff
@@ -95,11 +95,14 @@ it will create a new folder inside `packages` with a simple boilerplate for your
 - When pushing the release PR, use `--follow-tags` to make sure the relevant tag is also pushed
 - The CI will start doing the release when the tag is created
 - Don't forget to merge your PR
+- Release a new version of the documentation (see below)
```
I don't know what releasing a new version of the API docs will look like now that we have docusaurus (thoughts @smallsaucepan?).  I'm happy to wait for that to become clear in the Documentation section below first.

smallsaucepan replied:
SuffIcient to just update the turf-www CONTRIBUTING?

twelch replied:
Agreed, updated to point people to turf-www CONTRIBUTING to release a new version

twelch commented:
Appreciate the thoughts.   To answer your question @mfedderly, yes the auto release notes uses the PR title or commit description.  Here's an example of what they could look like:

# What's Changed
* Simplify unnecessary union type in turfDistance by @isti115 in [#2618](https://github.com/Turfjs/turf/pull/2618)
* Add inches, yards, meters to turf-distance units parameters by @twelch in [#346](https://github.com/Turfjs/turf/pull/346)
...[further list of commits]

**Full Changelog**: https://github.com/Turfjs/turf/compare/v6.5.0...v7.0.0

With the blessing you gave @mfedderly, as the person making most of the changelog edits lately, I'd like to try this auto-approach and just deprecate the changelog.  I've added to this PR doing so, and started to draft some  new instructions in the CONTRIBUTING doc.  I think there's still something missing here @mfedderly if you don't mind taking a look.

My observation is that people aren't really updating the changelog as we go.  Edits to it feel very ad-hoc right now, and not done as PR's come in, or just after, but rather right before a pre-release or release by a core contributor reviewing the commit log.   @mfedderly you were the main person doing changelog updates lately, so your blessing on switching to Github releases is well heard.

twelch commented:
Thanks @mfedderly I've updated the release instructions accordingly if you could give a final look.

twelch commented on CONTRIBUTING.md:
```diff
@@ -91,15 +91,54 @@ it will create a new folder inside `packages` with a simple boilerplate for your
 ### Prerelease
 
-- Every merged PR should trigger an automatic canary release
+- Every commit or PR merged to the master branch will trigger the Github [prerelease](https://github.com/Turfjs/turf/blob/master/.github/workflows/prerelease.yml) action, creating a canary release.
```
@mfedderly yes I think turning prerelease back on sounds great.

twelch replied:
Noted in docs that prerelease will be kept disabled but available.

smallsaucepan commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,8 +1,11 @@
 - [ ] Run `npm test` at the sub modules where changes have occurred.
+- [ ] Add tests for use cases addressed by the PR (if relevant)
+- [ ] Update JSDoc comments above changed functions (if relevant)
 - [ ] Run `npm run lint` to ensure code style at the turf module level.
```
The test task a few lines above should have linted the code, so no need to include this as a seperate task.

twelch replied:
@smallsaucepan are you saying the linter updates the JSDoc comments above functions?  The intention here is to cover a couple more essential the contributor should provide for a complete PR.

twelch replied:
My understanding is the linter only regenerates the readmes from the JSDoc comments.  It's up to the developer to update the JSDoc comments themselves.

twelch replied:
I updated the template text a little bit to clarify what is expected.

twelch replied:
I feel like more should be in the PR checklist, but it also should be simple.  I just noticed the "how to contribute" link referenced in bullet 3 actually doesn't quite exist anymore.  I'm going to shape that up a little bit to pick up the slack of what should be in a good PR.  This PR has spidered a little bit, but I think in a good way.

smallsaucepan replied:
> @smallsaucepan are you saying the linter updates the JSDoc comments above functions?

More that running "npm test" runs "npm run lint" automatically anyway.

twelch replied:
Got it.  I've now moved these steps under a single linked doc  This makes the pull request template much simpler.

- [ ] Have read and followed the steps for [preparing a pull request](https://github.com/Turfjs/turf/blob/master/CONTRIBUTING.md#preparing-a-pull-request).

smallsaucepan commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,8 +1,11 @@
 - [ ] Use a meaningful title for the pull request. Include the name of the package modified.
+- [ ] Is this just a bug fix, new functionality, or breaking change?
 - [ ] Have read [How To Contribute](https://github.com/Turfjs/turf/blob/master/CONTRIBUTING.md#how-to-contribute).
 - [ ] Run `npm test` at the sub modules where changes have occurred.
```
Should they be running test across the whole repo? The change may work in the sub module, but have a flow on effect in another package.

mfedderly replied:
It is safer to run them at the root, but I think we also have more trustworthy CI now that should hopefully catch anything that slips through.

twelch replied:
I've changed it to root, but made it clear package level is available for active development.

smallsaucepan commented on CONTRIBUTING.md:
````diff
@@ -91,15 +91,54 @@ it will create a new folder inside `packages` with a simple boilerplate for your
+
+- check the changed files and stage them for commit.  Example:
+```bash
+git add packages/*/package.json
````
lerna.json also gets updated right? Should that be committed at this stage as well @mfedderly?

mfedderly replied:
I was writing these instructions from memory so I probably just forgot that it changed, but yes everything the lerna command changes needs to get added.

twelch replied:
I've simplified the add command to stage all modifications.

twelch replied:
Actually I switched it to being more specific, avoid any surprises 🤷 

```bash
git add packages/*/package.json
git add lerna.json
```

smallsaucepan commented on CONTRIBUTING.md:
```diff
@@ -91,15 +91,54 @@ it will create a new folder inside `packages` with a simple boilerplate for your
+#### Release Final Steps
+- Create a Pull Request for the release, using the link in the output of the push command and the version number for the PR title (e.g. v7.0.0).  If the link isn't provided, just create a PR setup to merge your release branch to master.  Don't merge the PR yet.  Here is an example - https://github.com/Turfjs/turf/pull/2615
+
+- You can view the status of the triggered release action here - https://github.com/Turfjs/turf/actions.  Once complete, a new [version](https://www.npmjs.com/package/@turf/turf?activeTab=versions) of all turf packages will have been published on NPM.
```
This paragraph probably belongs at the end of the prev section after `push ... --follow-tags`. If I understand that's what kicks off the release action.

twelch replied:
Done.  I actually merged the two bullets together

smallsaucepan commented on CONTRIBUTING.md:
```diff
@@ -91,15 +91,54 @@ it will create a new folder inside `packages` with a simple boilerplate for your
+- As part of the release action, a draft Github release will have been created at https://github.com/Turfjs/turf/releases with an auto-generated changelog.  Edit and add to the release notes for readability and completeness, specifically noting any breaking changes.  Use past releases as a guide.  Be sure to "Save draft" each time, then ask for a review from other contributors.  Once ready, click ...
 
 ## Documentation
 
```
Maybe "Turf's documentation (README.md files) are updated automatically on commit, based on the JSDoc comments in the package files.

If you wish though, you can update them prior by using the following ..."

twelch replied:
Good point, I've noted this in the first release section.

twelch replied:
@smallsaucepan where is the documentation triggered on commit?  I'm not seeing it.

twelch replied:
Okay, I see it in monorepolint.config.mjs, which is triggered in the pre-commit

smallsaucepan commented on CONTRIBUTING.md:
```diff
@@ -91,15 +91,54 @@ it will create a new folder inside `packages` with a simple boilerplate for your
+- As part of the release action, a draft Github release will have been created at https://github.com/Turfjs/turf/releases with an auto-generated changelog.  Edit and add to the release notes for readability and completeness, specifically noting any breaking changes.  Use past releases as a guide.  Be sure to "Save draft" each time, then ask for a review from other contributors.  Once ready, click ...
 
 ## Documentation
 
```
In "Other Dependencies" below (sorry, github won't let me add a comment on that line) where it says "yarn" it should say "pnpm"

Should be able to remove "Yarn will need to be installed on your computer ..." as pnpm is included in every node install by default.

twelch replied:
Done

smallsaucepan commented on CONTRIBUTING.md:
```diff
@@ -91,25 +91,73 @@ it will create a new folder inside `packages` with a simple boilerplate for your
-- The CI will start doing the release when the tag is created
-- Don't forget to merge your PR
+
+A turf release is done from your local computer using a clean, up-to-date version of the turf master branch.
```
There are a mix of local and remote actions and commits, so this is too vague. "A turf release is initiated from your local computer, and then built and deployed remotely via a github actions."

twelch replied:
Good stuff. Will revisit this section soon

smallsaucepan commented on CONTRIBUTING.md:
```diff
@@ -91,25 +91,73 @@ it will create a new folder inside `packages` with a simple boilerplate for your
+  - If there are breaking changes, then it should be a major version bump, e.g. 7.x.x to 8.0.0.  This project follows [semantic versioning](https://semver.org/).
+- When running release commands below, replace `7.0.0` with the version number you are releasing.  Now, let's get started.
+
+- fetch the latest code from remote origin
```
Do you think there is the potential for confusion and that someone might try this in their development fork? Should we specify cloning a new local repo?

```bash
git clone git@github.com:Turfjs/turf.git turf-release
cd turf-release/
```

twelch replied:
Cloning a new local repo for a release seems like overkill (open to pushback).  But I understand the desire to make sure they have a fresh local copy of the right remote.  I am working on more specific steps for cloning and setting up your repo right now.  Let me take a shot at it.

smallsaucepan replied:
Fwiw, I wouldn't bat an eye at having to clone a fresh repo for a release. It takes one minute, 100MB of disk space, and eliminates any potential for confusion or cruft from previous builds. Worth the cost to benefit for something we're probably only doing once a month.

twelch replied:
Is there anything more to be done if working with a fresh install?  `pnpm install` at least?

smallsaucepan replied:
Lucky you got me to check. Yes, lerna needs nx installed, so it would be an `pnpm install`. Which builds the whole project unnecessarily, but I don't know a way around that at the moment.

twelch replied:
Thanks.  I'll move towards checking some of this

twelch replied:
I included the fresh clone steps as you suggested.  As a follow-on to that I included how to work with an existing copy, with extra cleanup/verify steps.

smallsaucepan commented on CONTRIBUTING.md:
````diff
@@ -91,25 +91,73 @@ it will create a new folder inside `packages` with a simple boilerplate for your
+
+- increment the version number of all packages, without pushing to origin.  This will also create a release tag.
+```bash
+pnpm lerna version --no-push 7.0.0
````
Should this include `--no-commit-hooks` as well? Otherwise we might accidentally generate readmes, etc. Ideally it shouldn't happen, but if something goes awry the release PR would end up containing things other than packages/*/package.json and lerna.json

twelch replied:
Added, seems logical.  Testing needed.

twelch commented:
Okay @smallsaucepan @mfedderly  I've reworked and simplified more of the Contributing page, filling in lots of holes for someone starting from scratch (and myself not having done a release).  I've simplified the pull request template to point to a checklist in a new "Preparing a pull request" section of the doc.

smallsaucepan commented:
I think this works out to around 1 comment per 4 lines of code changed, which is probably a new record. Nice work.
